Takanon common knowledge

noamoss opened the pull request:
closes #3 
closes #8

akariv commented on specs/takanon/config.yaml:
```diff
@@ -1,11 +1,10 @@
-  #   # TODO MAKE SURE THE FILE IS THERE
+context:
+  - name: ידע רלוונטי על התקנון
+    split: common-knowledge
```
what's the purpose of the split parameter here? It's conflicting with the source parameter.

noamoss replied:
The split parameter was intended to specify the local destination format for the downloaded source, but looking at the implementation, this might be redundant since we're using a hardcoded common-knowledge.md convention. I can remove the split parameter to avoid confusion.

akariv replied:
'split' was added as a different way to specify the source - a single markdown file with multiple entries (similar to the google spreadaheet - a single sheet with multiple entries).
These entries are then transformed to individual "files" in a vector store.

noamoss replied:
I know and thought to use is in a different way. Anyway, the recent changes the split is part of the vector store upoading and not for itself. It might be changed once we add other implementation other than OpenAI Vector Store

akariv replied:
i guess this now needs to change to `type: "spreadsheet"`?

noamoss replied:
fixed

noamoss commented on botnim/sync.py:
```diff
@@ -52,115 +53,68 @@ def openapi_to_tools(openapi_spec):
+        'instructions': config['instructions'],
+        'temperature': 0.00001,
+    }
+    kb_backend.set_assistant_params(asst_params)
```
The current naming is confusing. The set_assistant_params method isn't actually creating an assistant - it's just storing the configuration in memory for later use by the vector store backend. 

I'll rename it to store_assistant_config and move it after the assistant creation to make the flow clearer. The actual assistant creation only happens once, through the OpenAI client.

akariv replied:
1. I see at the moment calls to 'client.beta.assistants.create' both in openai.py and in 'sync.py'
2. Why would the vector store client need to know anything about the assistant in the first place? It is only supposed to care about vector stores, regardless of how they're used later.

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/10c06ad8f5fcbca14c949eb5a6b10cadd85f4c64

noamoss commented on botnim/kb/openai.py:
```diff
@@ -0,0 +1,215 @@
+
+client = OpenAI(api_key=api_key)
+
+class OpenAIVectorStore(KnowledgeBase):
```
You're right. 

I'll refactor this to separate the vector store operations from assistant management.

akariv replied:
Please keep the assistant creation code in sync.py at the moment.
We may refactor that in a future PR, but I want to focus on a single change here.

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/10c06ad8f5fcbca14c949eb5a6b10cadd85f4c64

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+        if 'split' in context_config:
+            split_docs = self._process_split_file(context_config)
+            if split_docs:
+                documents.extend(split_docs)
```
While currently we're using it with downloaded Google Spreadsheets, we might want to use this splitting functionality with local files or content from other sources in the future. That's why it's handled separately in the processing phase.

akariv replied:
I mean in this function - why not handle all three kinds of source in the same place:
- Existing separate markdown files (e.g. the takanon files)
- Existing single markdown file that needs to be split into separate "files" in the vector store (the "split" option)
- Rows in a google spreadsheet that transform to separate "files" in the vector store (the "source" option)

I think all three should be handled in the same place.

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/10c06ad8f5fcbca14c949eb5a6b10cadd85f4c64

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+            return []
+
+        documents = []
+        for file_path in sorted(dir_path.glob('*.md')):
```
You're right - there's a disconnect between where files are split (during download) and where they're processed. The split parameter was misleading since the actual splitting happens in download_sources.py. I'll refactor this to make the flow clearer by:
1. Removing the split parameter 
2. Renaming _process_split_file to better reflect what it does
3. Making the connection between downloading/splitting and processing more explicit in the code

akariv replied:
Why do we need to remove the split functionality at all? It might come handy in the future.

noamoss replied:
I asked my llm friend to update docstrings and logic. Please approve I got your intention correctly on this one.

akariv replied:
Please re-read the original code regarding the 'split' parameter and let me know how this is relevant...

noamoss replied:
we have a miscommunication here, I am afraid, let's discuss.
There are two 'splits' we are talking about: the parameter and the actual breaking down of sources.

akariv replied:
not sure how the `split` parameter would affect 'google spreadsheet' sources or `files` sources - for google spreadsheet sources the handling is always the same (each row to a chunk) and for files sources it's similar (one file per chunk).
Same goes for 'splitting', as implemented in the current code (I added a comment in the exact location for your reference) - it's only relevant when you have one file that you want to split into multiple chunks (as we do today with the `common-knowledge.md` file). 
What does 'file splitting' mean for any other source type, and how would that boolean parameter affect them?

noamoss replied:
> for google spreadsheet sources the handling is always the same (each row to a chunk)

this is an assumption, not a fact. In other words, we might want to push and analyze sheets/tables as a whole.

> What does 'file splitting' mean for any other source type

pages? semantic chunking?

noamoss commented on botnim/kb/base.py:
```diff
@@ -0,0 +1,41 @@
+
+    @abstractmethod
+    def exists(self, name: str) -> Tuple[bool, str]:
+        """Check if a knowledge base exists and return (exists, id)"""
```
Good catch.

noamoss replied:
It has been addressed by removing the method entirely as part of our separation of concerns refactoring. The assistant existence check is now handled directly in sync.py, and vector store operations are kept simple and focused.

noamoss commented on botnim/config.py:
```diff
@@ -1,7 +1,11 @@
 dotenv.load_dotenv(ROOT / '.env')
+
+# Logging configuration
+logging.basicConfig(level=logging.INFO)
```
+1

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/2fdce411b322149422751c129cc90fdaff31b256

noamoss commented on botnim/kb/base.py:
```diff
@@ -0,0 +1,41 @@
+from typing import List, Union, BinaryIO, Tuple
+import logging
+
+logger = logging.getLogger(__name__)
```
+1

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/2fdce411b322149422751c129cc90fdaff31b256

noamoss commented on botnim/cli.py:
```diff
@@ -21,17 +24,38 @@ def sync(environment, bots, replace_context=False):
+    click.echo(f"Downloading external sources for {bots}...")
+    
+    # First remove existing .md files or directories
+    if bots == 'all':
```
You're right. The hardcoded common-knowledge.md was a remnant from when we used the split parameter. I'll update the code to derive the paths from the configuration files themselves to determine which directories need to be cleaned before downloading.

noamoss commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,77 @@
+from pathlib import Path
+import os
+
+logger = logging.getLogger(__name__)
```
+1

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/2fdce411b322149422751c129cc90fdaff31b256

noamoss commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,77 @@
+        print(response.text[:500])  # Print first 500 chars to see the structure
+        
+        # Use csv module to properly handle quoted fields with commas
+        import csv
```
+1

noamoss commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,77 @@
+            # Add each non-empty column with its header
+            for header, value in zip(headers, columns):
+                if value:  # Only add non-empty values
+                    entry.append(f"{header}:\n{value}\n\n")  # Each field on new line for clarity
```
currently, to enable the contextual retrieval of common knowledge for a specific clause/section, but later to enable more metadata attached to each fact in the common knowledge.

noamoss commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,77 @@
+        if bot_filter != 'all' and bot_name != bot_filter:
+            continue
+            
+        import yaml
```
+1

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/0f0528a7d95fba0d7dda6e9d461e549d6e2dccc2

noamoss commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,77 @@
+            
+        if config.get('context'):
+            for context in config['context']:
+                if 'split' in context and 'source' in context:
```
as mentioned in the comment below : if a source has a split we are creating a named folder and putting all the parts (files) in it. Otherwise, the output will be just a single markdown file

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+        return name
+
+    def process_context(self, context_config: dict, replace: bool = False) -> Tuple[str, str]:
+        """Process a context configuration and return (vector_store_id, assistant_id)"""
```
you're right

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/d24048cbb6ed8195966ecfbad7c3f8164c8345d9

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+                logger.info(f"Deleting existing knowledge base: {kb_name}")
+                self.kb_backend.delete(assistant_id)
+                # Create new vector store and get new IDs
+                vector_store_id, assistant_id = self.kb_backend.create(kb_name)
```
now returns only the vector store ID. Assistant management has been moved to sync.py

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+        
+        return vector_store_id, assistant_id
+
+    def _process_files(self, file_pattern: str) -> List[BinaryIO]:
```
good catch. fixed

noamoss replied:
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/8a26e0144b1305c779896bc820ec61829b6bc5f1

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+        """Process regular files matching the pattern"""
+        files = sorted(self.config_dir.glob(file_pattern))
+        # Verify files have supported extensions
+        supported_extensions = {'.txt', '.md', '.pdf', '.doc', '.docx'}
```
> Do we actually support pdf's and doc's?

questionable: we are not splitting them, but as long as we are using openai api, we can upload and vectorize all these types. So my assumption is that in case of a real need we will also extend the splitting mechanism for these types. WDYT?

> Why do we return a 'text/plain' mimetype either way?

I thought there was a good reason for that - but seems I was wrong.

akariv replied:
Again, a separate PR for this new and unrelated functionality.

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+
+    def _process_split_file(self, context_config: dict) -> List[Tuple[str, BinaryIO, str]]:
+        """Process a directory of split files"""
+        dir_path = self.config_dir / context_config['split'].replace('.txt', '')
```
no good reason, a remaining of a previous try. removed.

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,90 @@
+            documents.extend(self._process_files(context_config['files']))
+
+        # Process split files (e.g., common knowledge)
+        if 'split' in context_config:
```
fixed, with explicit source types conditions/definitions

noamoss commented on README.md:
```diff
@@ -40,15 +45,49 @@ $ botnim --help
+5. Commit the changes to the repository
+6. Run the 'Sync' action from the GitHub Actions tab.
+
+### Knowledge Base Management
```
https://github.com/whiletrue-industries/rebuilding-bots/pull/9/commits/7744c8eac4d6385b1fe7aa21d9b3bcbe6d32cd13

noamoss replied:
oops. sorry, to be fixed.

noamoss commented:
@akariv ready for another iteration. Notice that now, with the applied changes, I am not saving local copies of the splitted common knowledge files. If/when we enable differential sync (of specific components) I will return it.

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,140 @@
+        elif source_type == 'spreadsheet':
+            # Process Google Spreadsheet source
+            if 'source' in context_config:
+                from .download_sources import download_and_convert_spreadsheet
```
import on top plz

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,140 @@
+                source_path = self.config_dir / context_config['source']
+                target_dir = source_path.parent / f"{source_path.stem}_split"
+                target_dir.mkdir(exist_ok=True)
+                self._split_file(source_path, target_dir)
```
function exists?

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,62 @@ def openapi_to_tools(openapi_spec):
-                    file_streams = [(f'{name}_{i}.md', f, 'text/markdown') for i, f in enumerate(file_streams)]
-                vector_store = client.beta.vector_stores.create(name=name)
-                while len(file_streams) > 0:
-                    file_batch = client.beta.vector_stores.file_batches.upload_and_poll(
```
upload_and_poll seems to be a more robust way of uploading the files to the vector store, at least for openai we saw that there were failures when uploading too many files at the same time.

noamoss replied:
done.

akariv replied:
hmm... what was done? I don't see it used anywhere?

noamoss replied:
I guess it was not committed before, sorry. fixed.

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,62 @@ def openapi_to_tools(openapi_spec):
-                        ...
-                    content = filename.read_text()
-                    content = content.split('\n---\n')
-                    file_streams = [io.BytesIO(c.strip().encode('utf-8')) for c in content]
```
working with BytesIO to hold the spreadsheet rows / split file parts is a much simpler and cleaner way to pass the information, instead of creating all these temporary files inside the repo (and having to delete them afterwards)

noamoss replied:
I Modified the upload system to handle both file paths (for regular files) and BytesIO objects (for spreadsheet data).
I thought we should keep the local cache files for spreadsheet data to enable incremental updates and change tracking

So now we are:
   - Avoiding unnecessary file I/O operations
   - Only uploading new or modified entries
   - Using in-memory operations where possible

akariv replied:
Let's move the cache functionality to a separate PR.
Either way:
1. I'm not sure if it's actually needed (i.e. what's the use case?). Remember that when running this normally on the server, it would not have any cache directory present, so it's mostly for dev purposes.
2. Not sure if this is the correct approach (i.e. actual separate files intermingled with the soruce code). A separate `.cache` folder or some other mechanism (e.g. `cache.pickle`) might work better IMO.
3. This is a clear case of functionality that should be separated to an external class.

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,140 @@
+            return f"{name_parts[0]} - פיתוח.{name_parts[1]}" if len(name_parts) > 1 else f"{name} - פיתוח"
+        return name
+
+    def process_context(self, context_config: dict) -> str:
```
This is a very misleading method name... as the context is not in fact processed but a vector store is created (regardless if an existing one already exists).

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,62 @@ def openapi_to_tools(openapi_spec):
-            name = context_['name']
-            if not production:
-                name += ' - פיתוח'
-            vector_store = client.beta.vector_stores.list()
```
This part also disappeared - what if I don't want to update the context?

noamoss replied:
to be fixed, with a better implementation.

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,62 @@ def openapi_to_tools(openapi_spec):
-
-    # List all the assistants in the organization:
-    assistants = client.beta.assistants.list()
+    """Update or create an assistant with the given configuration"""
```
Seems that 'replace_context' parameter is now completely ignored.

noamoss commented on botnim/cli.py:
```diff
@@ -39,19 +40,21 @@ def download(bots):
+        with config_file.open() as f:
+            config = yaml.safe_load(f)
+            
+        if config.get('context'):
```
I think we do, but you can explain the reason you are asking?

akariv replied:
It's related to another comment regarding using file streams instead of actual files.
Sorry this was a bit unclear.

noamoss replied:
already fixed, correct me if I am wrong.

noamoss commented:
@akariv ready for another round

akariv commented on botnim/cli.py:
```diff
@@ -13,13 +13,11 @@ def cli():
 @cli.command()
-@click.argument('environment', type=click.Choice(['production', 'staging']))
-@click.argument('bots', type=click.Choice(['budgetkey', 'takanon', 'all']))
-@click.option('--replace-context', is_flag=True, default=False, help='Replace existing context')
```
We should keep this functionality, though - not remove it.

noamoss replied:
@akariv 
we said it will be added on a separate task, once we move out of openai.

akariv replied:
No, we talked about selecting a specific file/source to update.
This parameter simply states whether we want to sync the agent and the context or just the agent.
Again - there is no need to remove functionality that's unrelated to the purpose of this PR (adding google spreadsheet as a source).

akariv commented on botnim/kb/download_sources.py:
```diff
@@ -5,51 +5,85 @@
         
-        print("Raw response text:")
-        print(response.text[:500])  # Print first 500 chars to see the structure
+        # Load existing files for comparison
```
Why do we need to download to actual files?

noamoss replied:
To support local actions (like exploring chunking, or prompt fine tunes). They can also act as a cache to detect changes in the source spreadsheet.

akariv replied:
Not sure how creating local files helps chunking or fine tunes (as you're going to load it to your local ES anyway...), but either way see my other comment on caching.

noamoss replied:
was added by mistake directly in the repository.

noamoss replied:
to be fixed.

akariv commented on specs/takanon/common-knowledge.md:
```diff
@@ -0,0 +1,7 @@
+"בניגוד לחוקים רגילים, התקנון לא מחייב אישור של נשיא המדינה או פרסום ברשומות כדי להיכנס לתוקף. זה מעניק לו מעמד ייחודי: הוא מסמך בעל תוקף משפטי פנימי שמסדיר את פעילות הכנסת, אך אינו כפוף ישירות לתהליכי החקיקה הסטנדרטיים."
```
Why is this here now?

noamoss replied:
was added by mistake directly in the repository.

akariv commented on specs/takanon/agent.txt:
```diff
@@ -1,31 +1,31 @@
-**You are a specialized AI agent trained to answer users' questions based on the Israeli parliament by-laws (תקנון הכנסת) and other approved legal texts in your knowledge base. Your priority is to answer using the by-laws when possible, and only supplement your answers with additional legal texts when necessary. If a question falls outside of these sources or the scope of your training, clearly s ...
+Your task is to answer users' questions based strictly on the attached provided legal texts concerning the Israeli parliament by-laws (תקנון הכנסת) and other specified legal documents. Do not use any other external sources to provide answers.
```
Not part of this pr.

noamoss replied:
You're right - I've reverted the agent instruction changes

noamoss replied:
You're right about the caching functionality - I'll move it to a separate PR.

akariv reviewed (changes requested):
There are still many unaddressed comments. If you want, let's have a call to discuss?

noamoss commented on botnim/sync.py:
```diff
@@ -52,115 +52,62 @@ def openapi_to_tools(openapi_spec):
-        instructions=config['instructions'],
-        temperature=0.00001,
-    )
-    if config.get('tools'):
```
>Where did this part go?

not sure what do you mean, if still relevant, please be more explicit.

> Note that in the original code we create/update the assistant after we know the vector store id, and update the tool list accordingly.

right, now the assistant is created first if it doesn't exist, and the context and tools are set up afterward (if necessary).

akariv replied:
See for example line 136 here with the content "if tool == 'code-interpreter':".
Where did this part go?

noamoss replied:
I did not see a reason to keep it, reverted.

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,106 @@ def openapi_to_tools(openapi_spec):
         for tool in config['tools']:
             if tool == 'code-interpreter':
-                tools.append(dict(type='code_interpreter'))
+                tools.append({"type": "code_interpreter"})
```
why this change?

noamoss replied:
just a style preference. reverted.

akariv replied:
was it though...?

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,106 @@ def openapi_to_tools(openapi_spec):
+                logger.warning(f'Error accessing existing vector store: {e}')
+                logger.warning(f'Assistant tool_resources: {existing_assistant.tool_resources}')
 
+    return assistant_id, vector_store_id
```
Why do we need to return this?

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,106 @@ def openapi_to_tools(openapi_spec):
+            )
+            logger.info(f'Assistant updated with new vector store: {assistant_id}')
+        else:
+            # Keep existing vector store connection
```
This is supposed to query the list of existing vector stores here and find an existing one, based on its name. This code right now does basically nothing...

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,106 @@ def openapi_to_tools(openapi_spec):
+            logger.info("Setting up new contexts...")
+            vector_store_id = context_manager.setup_contexts(config['context'])
+            
+            # Update assistant with new vector store
```
Again, this entire bit of code makes much more sends to happen *before* we start working on the assistant. The only thing we need to update are the `tools` and `tool_resources` and let the assistant code will handle the create/update sequence.

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,106 @@ def openapi_to_tools(openapi_spec):
+        )
+        logger.info(f'Updated assistant configuration: {assistant_id}')
+    
+    # Set up context if configured
```
do this first

akariv commented on botnim/sync.py:
```diff
@@ -52,115 +52,106 @@ def openapi_to_tools(openapi_spec):
+        logger.info(f"Context configuration found, replace_context={replace_context}")
+        if replace_context:
+            logger.info("Setting up new contexts...")
+            vector_store_id = context_manager.setup_contexts(config['context'])
```
this should return new entries for 'tools' and 'tool_resources' instead of the vector_store_id (which is an openai specific concept).
This will allow us to move to elasticsearch seamlessly.

akariv commented on botnim/kb/base.py:
```diff
@@ -0,0 +1,31 @@
+        """Upload documents to the vector store"""
+        pass
+
+    @abstractmethod
```
is this being used anywhere?

noamoss replied:
not yet, but I want to keep it a placeholder for future usage.

akariv commented on botnim/kb/base.py:
```diff
@@ -0,0 +1,31 @@
+
+logger = get_logger(__name__)
+
+class VectorStore(ABC):
```
I think that 'list' is missing here.

akariv replied:
(that is list vector stores)

noamoss replied:
fixed

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+            str: ID of the created vector store
+        """
+        kb_name = context_config['name']
+        vector_store_id = self.kb_backend.create(self._add_environment_suffix(kb_name))
```
check if exists?

noamoss replied:
added check

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+            return None
+        
+        # Create vector store using the first context's name
+        vector_store_id = self.create_vector_store(contexts[0])
```
The fact that this OpenAI specific behavior is here is a sign that there's a problem with the design...

This function should collect a list of lists of documents, then pass to the vectorstore implementation to create one or more stores in an implementation specific way.

noamoss replied:
done

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+                logger.warning(f"No documents found for context: {context.get('name', 'unnamed')}")
+        
+        if all_documents:
+            self.kb_backend.upload_documents(vector_store_id, all_documents)
```
this method would get the list of lists of documents and the configuration of the vector_stores and return the tools/tools_resources.

noamoss replied:
done.

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+        
+        # Handle spreadsheet source
+        if context_config.get('type') == 'spreadsheet' and 'source' in context_config:
+            from .download_sources import download_and_convert_spreadsheet
```
imports to top of file.

noamoss replied:
done

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+        # Handle spreadsheet source
+        if context_config.get('type') == 'spreadsheet' and 'source' in context_config:
+            from .download_sources import download_and_convert_spreadsheet
+            split_path = (
```
what is this?

noamoss replied:
remaining of the local saving. removed.

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+        Returns:
+            List of documents in the format required by the vector store
+        """
+        documents = []
```
no need for this variable, right?

noamoss replied:
right, fixed.

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+        
+        # Handle split files
+        elif 'split' in context_config:
+            documents.extend(self._process_split_file(context_config))
```
can simply return here without extend (same for all branches)

noamoss replied:
ok

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+        
+        return documents
+
+    def _process_directory(self, dir_path: Path) -> List[Tuple[str, str, str]]:
```
unused code?

noamoss replied:
fixed

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+            return f"{name_parts[0]} - פיתוח.{name_parts[1]}" if len(name_parts) > 1 else f"{name} - פיתוח"
+        return name
+
+    def create_vector_store(self, context_config: dict) -> str:
```
probably should be delegated to the kb_backend implementation to decide how many vector stores it supports (one or multiple)

noamoss replied:
done.

akariv commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+logger = get_logger(__name__)
+
+class ContextManager:
+    def __init__(self, config_dir: Path, kb_backend: VectorStore):
```
kb_backend -> vs_backend?

noamoss replied:
renamed

akariv commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,98 @@
+            content = ''.join(entry)
+            if content.strip():
+                filename = f"{context_name}_{i+1:03d}.md"
+                file_path = target_dir / filename
```
again, let's skip writing the actual files...

noamoss replied:
removed

akariv commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,98 @@
+        logger.debug(response.text[:500])
+        
+        # Create target directory if it doesn't exist
+        target_dir.mkdir(exist_ok=True, parents=True)
```
no need for target directory

noamoss replied:
removed

akariv commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,98 @@
+
+logger = get_logger(__name__)
+
+def download_and_convert_spreadsheet(source_url: str, target_dir: Path, context_name: str) -> List[Tuple[str, BinaryIO, str]]:
```
no need for target_dir

noamoss replied:
removed

akariv commented on botnim/kb/download_sources.py:
```diff
@@ -0,0 +1,98 @@
+        logger.error(f"Failed to download/convert source {source_url}: {str(e)}")
+        raise
+
+def download_sources(specs_dir: Path, bot_filter: str = 'all'):
```
I'm still not convinced we need this functionality in the CLI - can you please explain again?

noamoss replied:
removed

noamoss commented on botnim/kb/manager.py:
```diff
@@ -0,0 +1,179 @@
+        for context in contexts:
+            documents = self.collect_documents(context)
+            if documents:
+                all_documents.extend(documents)
```
done

akariv commented on botnim/sync.py:
```diff
@@ -52,121 +52,110 @@ def openapi_to_tools(openapi_spec):
-                            if ef.filename == f.name:
-                                client.files.delete(ef.id)
-                    file_streams = [f.open('rb') for f in files]
-                elif 'split' in context_:
```
See here the 'split' functionality - reading the source file and splitting it to bits that would be considered individual files.

noamoss replied:
yes, so I think the 'split' should be a true/false flag instead of serving two different purposes.

noamoss commented:
### Summary of Key Themes in the Pull Request

1. **Enhanced Knowledge Base Management:**
   - Introduced a modular system for managing knowledge bases with support for multiple backends (`OpenAI`, `files`, `spreadsheet`).
   - Added tools for handling context setup, document processing, and automated source downloads.

2. **Improved CLI Functionality:**
   - Extended `sync` command with options like `--replace-context` (rebuild vector stores) and `--debug` (debug mode).
   - Simplified CLI commands for better usability and flexibility.

3. **Refactored Project Structure:**
   - Modularized the codebase with new packages for vector store operations and context management (`vector_db`).
   - Enhanced maintainability through structured logging and streamlined configuration management.

4. **Configuration Updates:**
   - Updated bot configurations (`config.yaml`) to support new context types (e.g., spreadsheets) and improve flexibility.
   - Enabled dynamic source handling, including file splitting and external source integration.

5. **Asynchronous and Scalable Operations:**
   - Implemented asynchronous file uploads with progress tracking, batch processing, and retry mechanisms for better performance and error handling.

akariv commented on botnim/sync.py:
```diff
@@ -51,116 +52,87 @@ def openapi_to_tools(openapi_spec):
+    
+    # Set up context and get tools configuration
+    tools_config = None
+    if config.get('context') and replace_context:
```
If `replace_context` is not provided then the new/updated assistant will not have any vector stores attached.

akariv commented on botnim/sync.py:
```diff
@@ -51,116 +52,87 @@ def openapi_to_tools(openapi_spec):
+    return assistant_id
 
-def sync_agents(environment, bots, replace_context=False):
+async def sync_agents(environment, bots, replace_context=False, debug=False):
```
why are we making everything async now?

akariv commented:
Superseded by #20
